### PR TITLE
Eliminate vm.heapBaseAddress()

### PR DESCRIPTION
**Pull request by dchopra001**
This commit removes `J9::VMEnv::heapBaseAddress()` as this routine always returns 0. It was introduced in the codebase early on in the codebase and has since been superseded by a shift-based solution. Hence this is no longer needed. See eclipse/openj9#11263 for original issue and discussion.

Fixes: #11263

Signed-off-by: Dhruv Chopra <Dhruv.C.Chopra@ibm.com>

### CODE REVIEW

**Comment by dchopra001** [comment]
Related OMR PR: https://github.com/eclipse/omr/pull/5757

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp
Comment (fjeremic) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:487]: Can we fold uses of this as true?

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp
Comment (fjeremic) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:2640]: Similarly here.

File: runtime/compiler/codegen/J9CodeGenerator.cpp
Comment (fjeremic) [inline runtime/compiler/codegen/J9CodeGenerator.cpp:465]: Can we fold uses of this as true?

File: runtime/compiler/codegen/J9TreeEvaluator.cpp
Comment (fjeremic) [inline runtime/compiler/codegen/J9TreeEvaluator.cpp:502]: Can we fold uses of this as true?

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (fjeremic) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:1501]: Same here.

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (fjeremic) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:362]: Can we fold uses of this as true?

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (fjeremic) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp:9979]: Can we fold uses of this as true?

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp
Comment (fjeremic) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:485]: The code involving `translatedNode` seems to be dead, here and in all other codegens. Can we remove it?

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp
Comment (dchopra001) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:487]: Wouldn't that be incorrect? `usingCompressedPointers` is initialized as `false`. If we fold all uses to `true`, then once leaving this `if` block the code below may not generate the correct code under non-compressed refs:
```
TR::InstOpCode::Mnemonic storeOp = usingCompressedPointers ? TR::InstOpCode::strimmw : TR::InstOpCode::strimmx;
   TR::Register *translatedSrcReg = usingCompressedPointers ? cg->evaluate(node->getSecondChild()) : sourceRegister;
```

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp
Comment (dchopra001) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:2640]: Similar question as in https://github.com/eclipse/openj9/pull/11688/files#r560678679 about this one.

File: runtime/compiler/codegen/J9CodeGenerator.cpp
Comment (dchopra001) [reply runtime/compiler/codegen/J9CodeGenerator.cpp:465]: This is actually used later on in this method. It can potentially be redefined on line 637 below:
```
   static char *pEnv = feGetEnv("TR_disableLowHeapMem");
   if (pEnv)
      isLowMem = false;
```

This can influence whether certain branches are taken (not sure if we will ever use this environment variable option though? I don't think we will). Given all that, I've moved this definition closer to its first use for now. If you think it can be removed altogether then let me know and I'll do that.

File: runtime/compiler/codegen/J9TreeEvaluator.cpp
Comment (dchopra001) [reply runtime/compiler/codegen/J9TreeEvaluator.cpp:502]: I don't think there's any benefit to folding this one. `usingCompressedPointers` is the return value for this routine. It's initialized as `false` and only changed to `true` if the condition on line 466 above is met:
```
   if (cg->comp()->useCompressedPointers() && (node->getSymbolReference()->getSymbol()->getDataType() == TR::Address) &&
         (node->getSecondChild()->getDataType() != TR::Address))
```

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (dchopra001) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:362]: This is only `true` under compressedRefs on a 64-bit machine. We use this on line 463 below:
```
      if (loadNeedsDecompression)
         {
         if (comp->target().is64Bit() && comp->useCompressedPointers())

         if (isHB0)
            {
            if (shiftOffset > 0)
               {
               generateRegImmInstruction(SHL8RegImm1, node, loadOrStoreReg, shiftOffset, cg);
               }
            }
         else
            {
            TR_ASSERT(0, "!HB0 not supported yet");
            }
```

So I don't think we can fold this as true always.

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp
Comment (dchopra001) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp:9979]: Don't think so. This is initialized as `false`. If the function parameter `isObj` is false, then this boolean stays false as well. Folding it as `true` can lead to unwanted behaviour when it's used on line 9981 and 9993 below

File: runtime/compiler/codegen/J9CodeGenerator.cpp
Comment (fjeremic) [reply runtime/compiler/codegen/J9CodeGenerator.cpp]: I think we can remove it. There is not "disabling of low heap memory" anymore as it is always 0 now. So whatever code was dependent on that is now dead.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp
Comment (fjeremic) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:463]: I agree it "looks" wrong but it is what happens today. `usingCompressedPointers` is `true` in all paths of this function today. Whether that is correct or not is another question I suppose. Tagging @knn-k and @0xdaryl here for further thoughts. Do we have a bug here?

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (fjeremic) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:362]: All the paths it is used under is guarded with an `useCompressedPointers()` check though, so in effect all the locations where we query this values it is always true. Do you agree? Also the a name "HB0" no longer makes sense, and future readers will be quite confused as to what that even means.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp
Comment (fjeremic) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:463]: Yes, you're right. I made the mistake of missing the `if (comp->useCompressedPointers())` that this code is guarded by.

File: runtime/compiler/codegen/J9CodeGenerator.cpp
Comment (fjeremic) [inline runtime/compiler/codegen/J9CodeGenerator.cpp:635]: I think we should fold this as `true` and eliminate the variable along with the `feGetEnv` call. Note that `feGetEnv` will get called every time though this path now. This is not something we typically want as that call is quite expensive. This is why we typically make such calls and assign to `static` variables (so it only gets called once).

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (dchopra001) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:362]: The code is guarded by the `loadNeedsDecompression` boolean. This is defined as:
```
if (loadOrStoreOrArrayElementNode->getOpCodeValue() == TR::l2a || 
   ((loadOrStoreOrArrayElementNode->getOpCodeValue() == TR::aload || loadOrStoreOrArrayElementNode->getOpCodeValue() == TR::aRegLoad) && node->isSpineCheckWithArrayElementChild()) &&
       comp->target().is64Bit() && comp->useCompressedPointers())
      loadNeedsDecompression = true;
```

So if `loadOrStoreOrArrayElementNode->getOpCodeValue() == TR::l2a` is true, then `loadNeedsDecompression` will also be true, regardless of whether we are using compressed pointers, right?

File: runtime/compiler/codegen/J9CodeGenerator.cpp
Comment (dchopra001) [reply runtime/compiler/codegen/J9CodeGenerator.cpp]: 330ca59

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (dchopra001) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:362]: I don't think I explained my above comment sufficiently. What I meant is, we can end up in the following block:
```
      if (loadNeedsDecompression)
         {
         if (comp->target().is64Bit() && comp->useCompressedPointers())

         if (isHB0)
            {
            if (shiftOffset > 0)
               {
               generateRegImmInstruction(SHL8RegImm1, node, loadOrStoreReg, shiftOffset, cg);
               }
            }
         else
            {
            TR_ASSERT(0, "!HB0 not supported yet");
            }
```

Even if `useCompressedPointers()` is not true. So if we fold away `isHB0` as true, then we would indirectly assume that `loadNeedsDecompression` is also only ever true under compressed references. I agree though that the name `HB0` no longer makes sense. Perhaps we need to have a `bool is64BitCompressedRefs = comp->target().is64Bit() && comp->useCompressedPointers()` as this is used multiple times and can replace `isHB0` as well. Let me know what you think

Comment (fjeremic) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:362]: I'm not following why `loadNeedsDecompression` has anything to do with folding `isHB0`. The logic behind folding `isHB0` is simple:

- `isHB0` is set to true if and only if `TR::Compiler->vm.heapBaseAddress() == 0` **AND** `comp->target().is64Bit() && comp->useCompressedPointers()`
	- Heap base is always 0 so the former statement is always true
- Any uses of `isHB0` which are guarded by `comp->target().is64Bit() && comp->useCompressedPointers()` is therefore always true as well
	- There is only one other use of `isHB0` on line 472, and it is indeed guarded by `comp->target().is64Bit() && comp->useCompressedPointers(), therefore it can be folded as true

The logic follows that `isHB0` variable is true on all paths in which it is used, therefore we can just get rid of it. Does that make sense? Notice `loadNeedsDecompression` never came into the picture of such logic as whether this variable is true or not is irrelevant.

Comment (fjeremic) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:362]: What is slightly confusing is that the original code does not use braces. I'll paste the original code here:

```
      if (loadNeedsDecompression)
         {
         if (comp->target().is64Bit() && comp->useCompressedPointers())

         if (isHB0)
            {
            if (shiftOffset > 0)
               {
               generateRegImmInstruction(SHL8RegImm1, node, loadOrStoreReg, shiftOffset, cg);
               }
            }
         else
            {
            TR_ASSERT(0, "!HB0 not supported yet");
            }
```

And how I would have wrote it:
```
      if (loadNeedsDecompression)
         {
         if (comp->target().is64Bit() && comp->useCompressedPointers())
            {
            if (isHB0)
               {
               if (shiftOffset > 0)
                  {
                  generateRegImmInstruction(SHL8RegImm1, node, loadOrStoreReg, shiftOffset, cg);
                  }
               }
            else
               {
               TR_ASSERT(0, "!HB0 not supported yet");
               }
            }
```

Does the braces and indentation help to understand my above logic?

Comment (dchopra001) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:362]: Ah I see the issue. This line `if (comp->target().is64Bit() && comp->useCompressedPointers())` was removed from my branch because it saw it as redundant with `isHB0`. 

I think it's easiest if I just restore the original `if (comp->target().is64Bit() && comp->useCompressedPointers())` and remove `isHB0`.

Comment (dchopra001) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: 2b7638c96a0bf1c5e17d228ba8ab3eaada141f3d

**Review by fjeremic — APPROVED** [review]

**Comment by fjeremic** [comment]
Jenkins test sanity all jdk8,jdk11 depends eclipse/omr#5757

**Comment by fjeremic** [comment]
Jenkins test sanity s390x jdk11 depends eclipse/omr#5757

**Comment by fjeremic** [comment]
Linux s390x issues is due to #11795. Will wait for a good. Will wait for the OMR-Acceptance tonight before merging.

**Comment by fjeremic** [comment]
OMR-Acceptance is broken at the moment due to some CMake issue. Waiting for a fix on that before we can merge this.

**Comment by fjeremic** [comment]
https://github.com/eclipse/openj9-omr/compare/openj9...master
>There isn’t anything to compare.
>openj9 and master are identical.

We're in a good state right now. Going ahead with the merge.

**Pull request merged**
